Inverting expressions containing Abs terms can returns the associated conditions

kshitij10496 opened the pull request:
As it was discussed #10532 , a ConditionSet (with all the necessary conditions) must be returned whenever applicable to represent the complete solution of the given equation.
A `FiniteSet` of solutions is returned whenever all conditions are satisfied else 
a `ConditionSet` object with all necessary conditions is returned.
The solutions present in the `ConditionSet` can be accessed by using its `base_set` attribute. 

For example:
**From current master**

``` python
In [ ]: x = symbols('x')
In [ ]: n = symbols('n')
In [ ]: solveset(abs(x)-n,x, S.Reals)
Out[ ]: Intersection([0, oo), {n})
```

**New implementation**

``` python
In [ ]: x = Symbol('x')
In [ ]: n = Symbol('n')

In [ ]: solveset(abs(x)-2, x, S.Reals)
Out[ ]: {-2, 2}

In [ ]: solveset(abs(x)-n, x, S.Reals)
Out[ ]: ConditionSet(x, Contains(n, [0, oo)), {-n, n})

In [ ]: solveset(abs(x)-n, x, S.Reals).base_set
Out[ ]: {-n, n}
```
- [x] closes #10532 
- [x] modified tests
- [x] modified docs

Ping @aktech @asmeurer @hargup @jksuom @smichr  
Kindly review this and let me know where I can make improvements. 

Reference: #10532 , #9588

smichr commented on doc/src/modules/solvers/solveset.rst:
```diff
@@ -398,7 +398,7 @@ How does ``solveset`` ensure that it is not returning any wrong solution?
     >>> x, n = symbols('x, n')
     >>> pprint(solveset(abs(x) - n, x, domain=S.Reals), use_unicode=True)
-    [0, ∞) ∩ {n}
+    {x | x ∊ {-n, n} ∧ ({{n} ∈ [0, ∞)})}
```
Is the rhs correct? It seems like it should read "^ {n \elof [0, oo)}".

kshitij10496 replied:
Actually, I did this is by design.
For example, if we have more than one condition to be satisfied as in this case [[Link](https://github.com/sympy/sympy/pull/10751/files#diff-85baa04bbf4e1dfd9128782738e45424R132)],
I think this representation will be better for handling various conditions.
What do you think ?

asmeurer replied:
I don't follow what this means. For starters, it seems there are redundant parentheses (is there a bug in the printer?), but removing those, we get `x ∊ {-n, n} ∧ {...}`. The `{...}` bit is a set, which isn't a boolean predicate, so it doesn't make sense to have it in an `And`. Finally, `{n}` is a set whereas `[0, ∞)` is a set of numbers, so it will never be an element of it. You probably meant to use "subset", not "element of". Do we have a `Subset` object? 

The cleanest way to write this would be to use ForAll, but we don't have that yet (see https://github.com/sympy/sympy/pull/7608). A Subset would also work (it would just be less readable), as would an And over all the elements (which would be longer, but also less roundabout). 

Also, `∊` and `∈` are different characters (the former is "contains" and the latter is an epsilon).

kshitij10496 replied:
- Yes.  The are some redundant parantheses here.
I wrapped the `Contains` object with a `FiniteSet` so as to facilitate Union operation on more `Contains` objects if there be such a situation. This contributes to an overhead curly brace of the output. We can mitigate this by adding more than one condition to the `ConditionSet` object.
Would passing an `And` object here be a good solution to this issue ?
-  I used a FiniteSet object to represent the conditioned symbols so as to give a compact output.
I was having an issue with using the args of the `FiniteSet`
For example:
When using the `FiniteSet.args`
Output : ` (n,) ∈ [0, ∞)`
Instead of `(n) ∈ [0, ∞) `  
Also, I want to know which of the following results should we prefer ?
`((b, a - b, a + b) ∈ [0, ∞))` or `(b, a - b, a + b ∈ [0, ∞)) or ` (b ∈ [0, ∞)) ∧ (a - b ∈ [0, ∞)) ∧ (a + b ∈ [0, ∞))`
- The `∈` (epsilon) in the output is due to the printer of `Contains` object.
I think this might be a bug.  

@asmeurer Thanks for this.
I Hope, I have

kshitij10496 replied:
I have updated the conditions to be a boolean predicate instead of being a set.
I would like to know which output is preferred so as to modify the set of conditioned symbols( or expressions).

Ping @aktech @asmeurer @hargup @smichr

Commit:
added _invert_abs function for inverting abs functions

kshitij10496 commented:
I think we restart the build now.

smichr commented:
Looks promising so far. What do you think about moving the code to a helper `_solve_abs` or similar that mimics other helpers?

kshitij10496 commented:
I think, now, `_solve_abs` primarily checks only for the `domain` argument of the equation containing `Abs` while the `invert_real` takes care of the solution.
Doesn't this change render `_solve_abs` redundant?
If yes, then we should move the code to a new helper, or better yet, replace the current `_solve_abs` with this.
What do you think about this ?

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -125,11 +126,62 @@ def _invert_real(f, g_ys, symbol):
+
+        if f.is_positive:
+            f_n = f
+        else:
```
Abs is always nonnegative so there is no need for an else, is there?

smichr replied:
Perhaps you meant to check the argument and are trying to make things simpler in terms of what is reported as having to be positive?

kshitij10496 replied:
> Abs is always nonnegative so there is no need for an else, is there?

I did this because being non-negative doesn't imply a positive value.
And the `else` condition takes care of that case.

> Perhaps you meant to check the argument and are trying to make things simpler in terms of what is reported as having to be positive?

Yes. Thats exactly what I am trying to do here.

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -125,11 +126,62 @@ def _invert_real(f, g_ys, symbol):
+        g_x, values = _invert_real(f_n, Union(imageset(Lambda(n, n), g_ys),
+                                   imageset(Lambda(n, -n), g_ys)), symbol)
+
+        if isinstance(values, ConditionSet):
```
OK, looking more closely I see that much of what you have here is just simplification of ConditionSet. That should probably be in ConditionSet (maybe _eval?).

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -125,11 +126,62 @@ def _invert_real(f, g_ys, symbol):
+                pass  # No possible usecase
+            values = values.base_set
+
+        if conditions == S.EmptySet:
```
nit: compare singletons with `is`

smichr replied:
This part here applies only to inverting Abs, so shouldn't it be in the `_solve_abs` routine?

kshitij10496 replied:
>  compare singletons with is
Thanks. I will make a note of this.

smichr commented:
OK, I see the confusion: there is a `_solve_abs` to solve an expression with 1 or more abs and then there is the code you have added. I think what you might want to do is make your code an `_invert_abs` routine which gets called from `_invert_real`. So basically, if it's not trivial to invert then invert calls its helper.

kshitij10496 commented:
This sounds cool.
I will make the changes

kshitij10496 commented:
Ping @smichr

hargup commented on sympy/solvers/solveset.py:
```diff
@@ -234,6 +231,69 @@ def _invert_complex(f, g_ys, symbol):
 
 
+def _invert_abs(f, g_ys, symbol):
+    """Helper function for inverting absolute value functions"""
```
In the docstring you should briefly explain how this sub routine works, this is not required for other `_invert_*` because they are largely trivial or the code is self explanatory.

kshitij10496 replied:
I needed a review of this as I was not sure about what should go in the docs.
Thanks.

Commit:
Added docstrings for the new _invert_abs function

kshitij10496 commented on sympy/solvers/solveset.py:
```diff
@@ -234,6 +231,84 @@ def _invert_complex(f, g_ys, symbol):
+    `EmptySet` is returned.
+    In all other cases, a `ConditionSet` of the solutions, with all the required
+    conditions specified, is returned.
+
```
Do we need to add examples too?
@hargup @smichr

jksuom commented on sympy/solvers/solveset.py:
```diff
@@ -234,6 +231,84 @@ def _invert_complex(f, g_ys, symbol):
+        f_n = f
+    else:
+        f_n = f.args[0]
+    g_x, values = _invert_real(f_n, Union(imageset(Lambda(n, n), g_ys),
```
What is the difference between ``imageset(Lambda(n, n), g_ys)`` and ``g_ys``? They should have the same elements.

kshitij10496 replied:
+1.
I wrote this because of two reasons: 

- To clearly show the logic: Inverting the `Union` of both the positive and negative images of `g_ys` under the transformation `f_n`.
- For code readability

smichr replied:
The IdentityFunction is detected and `g_ys` is returned, so little harm is done by doing this and the logic of what is being done (I agree) is made clear.

```python
>>> imageset(n,n,S.Integers)
Integers()
```

kshitij10496 commented:
Ping @aktech @asmeurer @hargup @jksuom @smichr

This is ready for review.
I hope to have addressed all the concerns above.
I would love to discuss any other issues with this.
Looking forward to your comments.

Commit:
updated _invert_abs with new solveset API

Commit:
updated tests for _invert_abs

Commit:
added XFAILing solvify tests

Commit:
added XFAIL tests for Abs functions using nonlinsolve

kshitij10496 commented:
@smichr Can you review the PR and let me know what do you think about the relevance of these changes now?

smichr commented:
Can you resolve the conflicts?

Commit:
Merge branch 'master' into abs_invert

smichr commented:
from Travis:

```
File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.1.2.dev0-py2.7.egg/sympy/utilities/runtests.py", line 1270, in _timeout
    function()
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.1.2.dev0-py2.7.egg/sympy/solvers/tests/test_solveset.py", line 60, in test_invert_real
    conditions = Contains(n, Interval(0, oo), evaluate=False)
NameError: global name 'Contains' is not defined
```
